c++/src/kj/common.h

```diff
@@ -895,6 +895,7 @@ class NullableValue {
 #endif
   {
     if (isSet) {
+      isSet = false;
```

Comment:
I suspect the compiler may be allowed to optimize away this assignment, since it can observe that once the destructor completes, the object has become uninitialized memory and its contents may as well be anything. So I think the refactoring you did here may not be safe, unfortunately.

Can we reduce this PR to just add the operator=, which should do `dtor(value)` instead of `dtor(*this)`?

Reply:
Ugh. I don't know if you're right, but that sounds like something C++ could do indeed. I'll revert it.